wasi/wasi.go

```diff
@@ -100,88 +170,102 @@ type fileEntry struct {
+	if len(args) == 0 {
+		return &nullTerminatedStrings{nullTerminatedValues: [][]byte{}}, nil
 	}
-	if len(args) > math.MaxUint32 {
-		return nil, fmt.Errorf("the length of the args exceeds the max of uint32: %v", len(args))
-	}
-	strings := make([][]byte, len(args))
+	var strings [][]byte // don't pre-allocate as this function is size bound
 	totalBufSize := uint32(0)
 	for i, arg := range args {
-		argLen := uint64(len(arg)) + 1 // + 1 for "\x00"
-		if argLen > uint64(math.MaxUint32-totalBufSize) {
-			return nil, fmt.Errorf("the required buffer size for the args exceeds the max of uint32: %v", uint64(totalBufSize)+argLen)
+		if !utf8.ValidString(arg) {
+			return nil, fmt.Errorf("arg[%d] is not a valid UTF-8 string", i)
+		}
+		argLen := uint64(len(arg)) + 1 // + 1 for "\x00"; uint64 in case this one arg is huge
+		nextSize := uint64(totalBufSize) + argLen
+		if nextSize > uint64(maxBufSize) { //
+			return nil, fmt.Errorf("arg[%d] will exceed max buffer size %d", i, maxBufSize)
 		}
-		totalBufSize += uint32(argLen)
-		strings[i] = make([]byte, argLen)
-		copy(strings[i], arg)
+		totalBufSize = uint32(nextSize)
+		strings = append(strings, append([]byte(arg), 0))
 	}
-
-	return &wasiStringArray{nullTerminatedValues: strings, totalBufSize: totalBufSize}, nil
+	return &nullTerminatedStrings{nullTerminatedValues: strings, totalBufSize: totalBufSize}, nil
 }
 
-// Args returns an option to give a command-line arguments to the WASIEnvironment.
-// Args returns an error if the length or the total size of the given string slice exceeds the max of uint32.
-func Args(args []string) (Option, error) {
-	wasiStrings, err := newWASIStringArray(args)
+// Args returns an option to give a command-line arguments to the API or errs if the inputs are too large.
+//
+// Note: The only reason to set this is to control what's written by API.ArgsSizesGet and API.ArgsGet
+// Note: While similar in structure to os.Args, this controls what's visible in Wasm (ex the WASI function "_start").
+func Args(args ...string) (Option, error) {
+	wasiStrings, err := newNullTerminatedStrings(math.MaxUint32, args...) // TODO: this is crazy high even if spec allows it
 	if err != nil {
 		return nil, err
 	}
-	return func(w *WASIEnvironment) {
+	return func(w *api) {
 		w.args = wasiStrings
 	}, nil
 }
 
 func Preopen(dir string, fileSys FS) Option {
-	return func(w *WASIEnvironment) {
+	return func(w *api) {
 		w.opened[uint32(len(w.opened))+3] = fileEntry{
 			path:    dir,
 			fileSys: fileSys,
 		}
 	}
 }
 
-func NewEnvironment(opts ...Option) *WASIEnvironment {
-	ret := &WASIEnvironment{
-		args:   &wasiStringArray{},
+// RegisterAPI adds each function API to the wasm.Store via AddHostFunction.
+func RegisterAPI(store *wasm.Store, opts ...Option) error {
+	_, err := registerAPI(store, opts...)
+	return err
+}
+
+// TODO: we can't export a return with API until we figure out how to give users a wasm.HostFunctionCallContext
```

Comment:
hmm maybe it is ok, just we need an example. Ex if you define your own host function, invocations to that will have an instance of wasm.HostFunctionCallContext, which you could pass to API. In other words, `wasi.API` would only be used inside something that already has a context (apart from these tests). Make sense?

Reply:
oh maybe I was wrong this doesn't work  for us

Reply:
nvm

Reply:
well I was right:

```

type API interface {
	Foo()
}

type wazeroImpl struct{}

func (a *wazeroImpl) Foo() {
	fmt.Println("wazeroImpl")
}

func newAPI() API {
	return &wazeroImpl{}
}

type userImpl struct {
	defaultImpl API // Embed the interface
}

func (i *userImpl) Foo() {
	i.defaultImpl.Foo()
	fmt.Println("userImpl")
}

func callFoo(impl API) {
	v := reflect.ValueOf(impl.Foo)
	v.Call(nil)
}

func Test_a(t *testing.T) {
	ownImpl := userImpl{newAPI()}
	callFoo(&ownImpl)
}
```

Reply:
as long as we have an API something like `RegisterOverriddenAPI(userImpl, store)` we could allow users to override the WASI func as much as they like __in theory__

Reply:
anyway LGTM for now

Reply:
cool I think we are talking about mildly different topics, but agree this change is probably a good unit to start with and also help unblock @nullpo-head while we think more